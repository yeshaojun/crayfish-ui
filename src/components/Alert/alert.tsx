import { FC, ReactNode, useState } from "react";
import classNames from "classnames";
export type AlertType = "success" | "info" | "warn" | "error";

export interface AlertProps {
  className?: string;
  closeable?: Boolean; // 是否显示close
  message: String;
  type?: AlertType;
  description?: string;
  showIcon?: Boolean;
  closeText?: string;
  icon?: ReactNode;
  afterClose?: () => void;
}

export const Alert: FC<AlertProps> = (props) => {
  const {
    closeable,
    message,
    type,
    description,
    className,
    showIcon,
    closeText,
    icon,
    afterClose,
    ...restProps
  } = props;
  const [closed, setClosed] = useState(false);

  const classes = classNames("alert", className, {
    [`alert-${type}`]: type,
  });

  const renderIcon = () => {
    if (showIcon) {
      if (icon) {
        return icon;
      } else {
        return type;
      }
    }
  };

  const close = () => {
    setClosed(true);
    afterClose && afterClose();
  };

  // 这里应当加入动画
  // antd使用CSSMotion方式，rc-motion库
  if (!closed) {
    return (
      <div className={classes} {...restProps}>
        {renderIcon()}
        <div className="alert-content">
          <div className="alert-content-message">{message}</div>
          <div className="alert-description">{description}</div>
        </div>
        <div className="alert-close-icon" onClick={close}>
          关闭
        </div>
      </div>
    );
  } else {
    return null;
  }
};

Alert.defaultProps = {
  closeable: false,
  showIcon: false,
  type: "success",
};
