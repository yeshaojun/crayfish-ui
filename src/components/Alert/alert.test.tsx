import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Alert, AlertProps } from "./alert";

const defaultProps: AlertProps = {
  message: "nihao",
};

describe("render alert component", () => {
  it("render default alert", () => {
    const wrapper = render(<Alert {...defaultProps}></Alert>);
    const element = wrapper.getByText("nihao");
    expect(element).toBeInTheDocument();
    expect(wrapper.container.tagName).toEqual("DIV");
    expect(wrapper.container.children[0]).toHaveClass("alert alert-success");
  });

  it("sets data attributes on input", () => {
    const { container } = render(
      <Alert data-test="test-id" data-id="12345" message="" />
    );
    const input = container.querySelector(".alert")!;
    expect(input.getAttribute("data-test")).toBe("test-id");
    expect(input.getAttribute("data-id")).toBe("12345");
  });

  it("support closeIcon", () => {
    const { container } = render(
      <Alert
        closeable
        message="Warning Text Warning Text Warning TextW arning Text Warning Text Warning TextWarning Text"
        type="warn"
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("could be closed ", () => {
    const onClose = jest.fn();
    const { container } = render(
      <Alert
        closeable
        message="Warning Text Warning Text Warning TextW arning Text Warning Text Warning TextWarning Text"
        type="warn"
        afterClose={onClose}
      />
    );
    fireEvent.click(container.querySelector(".alert-close-icon")!);
    expect(onClose).toHaveBeenCalled();
  });
});
