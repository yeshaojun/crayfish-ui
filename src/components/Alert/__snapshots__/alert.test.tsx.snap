// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render alert component support closeIcon 1`] = `
<div>
  <div
    class="alert alert-warn zoom-in-top-appear zoom-in-top-appear-active"
  >
    <div
      class="alert-content"
    >
      <div
        class="alert-content-message"
      >
        Warning Text Warning Text Warning TextW arning Text Warning Text Warning TextWarning Text
      </div>
      <div
        class="alert-description"
      />
    </div>
    <div
      class="alert-close-icon"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-xmark crayfish-icon icon-primary"
        data-icon="xmark"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M310.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L160 210.7 54.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L114.7 256 9.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L160 301.3 265.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L205.3 256 310.6 150.6z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
